packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/remote_constraint_provider.dart

```diff
@@ -22,20 +31,128 @@ abstract class RemoteLoggingConstraintProvider {
   DefaultRemoteLoggingConstraintProvider({
     required DefaultRemoteConfiguration config,
     required AWSCredentialsProvider credentialsProvider,
   })  : _config = config,
-        _credentialsProvider = credentialsProvider;
+        _credentialsProvider = credentialsProvider {
+    _fetchAndCacheConstraintFromEndpoint(); // Fetch once immediately on initialize
+    _refreshConstraintPeriodically();
+  }
 
   final DefaultRemoteConfiguration _config;
   final AWSCredentialsProvider _credentialsProvider;
 
+  LoggingConstraint? _loggingConstraint;
+
+  /// The signer to sign the request.
+  static const _signer = AWSSigV4Signer();
+
+  // The timer to refresh the constraint periodically.
+  Timer? _timer;
+
+  /// Whether the periodic fetch is running.
+  bool isRunning = false;
+
+  /// Stops the periodic fetch.
+  void stop() {
+    _timer?.cancel();
+    isRunning = false;
+  }
+
   @override
-  // TODO(nikahsn): add implementation.
-  LoggingConstraint get loggingConstraint => throw UnimplementedError();
+  String get runtimeTypeName => 'DefaultRemoteLoggingConstraintProvider';
+
+  Future<void> _saveConstraintLocally(LoggingConstraint constraint) async {
+    await storage.saveConstraintLocally(constraint.toJson());
+  }
+
+  Future<void> _fetchAndCacheConstraintFromEndpoint() async {
+    try {
+      final constraint = await _fetchConstraintFromEndpoint();
+      if (constraint != null) {
+        _loggingConstraint = constraint;
+        await _saveConstraintLocally(constraint);
+      }
+      // ignore: avoid_catches_without_on_clauses
+    } catch (error) {
+      logger.error('Error fetching constraints: $error');
+    }
+  }
+
+  Future<LoggingConstraint?> _fetchConstraintFromEndpoint() async {
+    final uri = Uri.parse(_config.endpoint);
+
+    final request = AWSHttpRequest.get(uri);
+
+    final scope = AWSCredentialScope(
+      region: _config.region,
+      service: AWSService.apiGatewayManagementApi,
+    );
+
+    final signedRequest = await _signer.sign(
+      request,
+      credentialScope: scope,
+    );
+
+    final headers = signedRequest.headers.map<String, String>(
+      MapEntry<String, String>.new,
+    );
+
+    final response = await http.get(
+      uri,
+      headers: {
+        ...headers,
+        HttpHeaders.acceptHeader: 'application/json; charset=utf-8',
+      },
+    );
+
+    if (response.statusCode == 200) {
+      final fetchedConstraint = LoggingConstraint.fromJson(
+        jsonDecode(response.body) as Map<String, dynamic>,
+      );
+      return fetchedConstraint;
+    }
+    throw Exception(
+      'Failed to fetch logging constraint from ${_config.endpoint}',
+    );
+  }
+
+  @override
+  LoggingConstraint? get loggingConstraint {
+    if (_loggingConstraint != null) {
+      return _loggingConstraint;
+    }
+    // Load from local storage synchronously
+    final file = File('logging_constraint.json');
+    if (file.existsSync()) {
```

Comment:
Should not use sync file methods basically ever on the main thread outside of CLI applications where it's totally okay to block